extern/test_bmi_cpp/include/test_bmi_cpp.hpp

```diff
@@ -0,0 +1,196 @@
+        int input_var_grids[INPUT_VAR_NAME_COUNT] = { 1, 1 };
+        int output_var_grids[OUTPUT_VAR_NAME_COUNT] = { 1, 1 };
+
+        std::map<std::string,int> type_sizes = {
+            {BMI_TYPE_NAME_DOUBLE, sizeof(double)},
+            {BMI_TYPE_NAME_FLOAT, sizeof(float)},
+            {BMI_TYPE_NAME_INT, sizeof(int)},
+            {BMI_TYPE_NAME_SHORT, sizeof(short)},
+            {BMI_TYPE_NAME_LONG, sizeof(long)}
+        };
+
+        // ***********************************************************
+        // ***************** Non-dynamic allocations *****************
+        // ***********************************************************
+
+        // Epoch-based start time (BMI start time is considered 0.0)
+        long epoch_start_time = 0.0;
+        int num_time_steps = 0;
+        double current_model_time = 0.0;
+        double model_end_time = 0.0;
+        int time_step_size = DEFAULT_TIME_STEP_SIZE;
+
+        // ***********************************************************
+        // ******************* Dynamic allocations *******************
+        // ***********************************************************
+        std::unique_ptr<double> input_var_1 = nullptr;
+        std::unique_ptr<double> input_var_2 = nullptr;
+        std::unique_ptr<double> output_var_1 = nullptr;
+        std::unique_ptr<double> output_var_2 = nullptr;
+
+        /**
+        * Read the BMI initialization config file and use its contents to set the state of the model.
+        *
+        * @param config_file The path to the config file.
+        * @param model Pointer to the model struct instance.
+        * @return The BMI return code indicating success or failure as appopriate.
+        */
+        void read_init_config(std::string config_file);
+
+        /**
+         * Read number of lines in file and max line length, returning -1 if it does not exist or could not be read.
+         *
+         * @param file_name The name of the file to open and read.
+         * @param line_count A pointer to a location in which to write the value for the number of lines in the file.
+         * @param max_line_length A pointer to a location in which to write the value of the max line length for the file.
+         * @return 0 if successful or -1 otherwise.
+         */
+        void read_file_line_counts(std::string file_name, int* line_count, int* max_line_length);
+
+        /**
+         * Run this model into the future.
+         *
+         * @param model The model struct instance.
+         * @param dt The number of seconds into the future to advance the model.
+         * @return 0 if successful or 1 otherwise.
+         */
+        void run(long dt);
+
+};
+
+extern "C"
+{
+    /**
+    * Construct this BMI instance as a normal C++ object, to be returned to the framework.
+    *
+    * @return A pointer to the newly allocated instance.
+    */
+	TestBmiCpp *bmi_model_create()
```

Comment:
As mentioned above, should the types for these functions be the base `bmi::Bmi` type instead?